### PR TITLE
Remove visualization member from visualizations controllers

### LINKED ISSUE #12185: [Model refactor] Remove CartoDB::Visualization::Member and Collection from *::VisualizationsController
Remove last references to old visualization models from visualizations controllers.

### PR DESCRIPTION
**Pull request by juanignaciosl**
This closes #12185.

Acceptance:

* Be careful with visualization location change at `#update`.

After each import/creation check the following pain points:
- Name.
- Permission.
- Data is rendered.
- No JS errors.

# With Editor:

- [x] Enable private tables and maps for the user. Import a map. Map should be private. Go to `.../public_map`. It should return an embed error page ("Looks like this map is set as private or no longer exists") with status code 403.
- [x] Import a dataset with drag and drop.
- [x] Create dataset from query.
- [x] Create empty dataset.
- [x] Import a dataset from data library and then delete it.
- [x] Create map from table view.
- [x] Create map with two data library datasets.
- [x] Import a map with drag and drop.
- [x] Duplicate a map
- [x] Create empty map.
- [x] Create map from sync.

*Edit a dataset*

- [x] Edit metadata.
- [x] Rename, with spaces. Spaces should be replaces with underscores (`_`).
- [x] Change privacy
- [x] Share with someone in the org

*Edit a map* (edit and check that changes are propagated to the public view):

- [x] Change style.
- [x] Enable layer selector.
- [x] Add one layer.
- [x] Edit metadata (description and tags).
- [x] Rename
- [x] Lock it.
- [x] Unlock it.
- [x] Change privacy
- [x] Share with someone in the org

# With Builder (same, copy & paste):

- [x] Enable private tables and maps for the user. Import a map. Map should be private. Go to `.../public_map`. It should return an embed error page ("Looks like this map is set as private or no longer exists") with status code 403.
- [x] Import a dataset with drag and drop.
- [x] Create dataset from query.
- [x] Create empty dataset.
- [x] Import a dataset from data library and then delete it.
- [ ] Create map from table view.
- [x] Create map with two data library datasets.
- [x] Import a map with drag and drop.
- [x] Duplicate a map
- [x] Create empty map.
- [x] Create map from sync.

*Edit a dataset*

- [x] Edit metadata.
- [x] Rename, with spaces. Spaces should be replaces with underscores (`_`).
- [x] Change privacy
- [x] Share with someone in the org

*Edit a map* (edit and check that changes are propagated to the public view):

- [x] Change style.
- [x] Enable layer selector.
- [x] Add one layer.
- [x] Edit metadata (description and tags).
- [x] Rename
- [x] Lock it.
- [x] Unlock it.
- [x] Change privacy
- [x] Share with someone in the org

### CODE REVIEW

**Comment by juanignaciosl** [comment]
Looking for a first CR. Key points: extra 👀  on risks (acceptance criteria is still empty, but it will contain a lot of editor/builder/permissions checks), and on TODOs (which to address, which not). Not looking for acceptance yet, only a first round of discussion.

cc @CartoDB/builder-backend

File: app/controllers/api/json/visualizations_controller.rb [hunk 1]
```diff
@@ -27,142 +23,6 @@ class Api::Json::VisualizationsController < Api::ApplicationController
...
-        CartoDB::Logger.error(message: "InvalidMember updating visualization", visualization_id: vis.id, exception: e)
-        render_jsonp({ errors: vis.full_errors.empty? ? ['Error saving data'] : vis.full_errors }, 400)
-      rescue => e
-        CartoDB::Logger.error(message: "Error updating visualization", visualization_id: vis.id, exception: e)
-        render_jsonp({ errors: ['Unknown error'] }, 400)
-      end
-    end
-  end
-
   def notify_watching
```
Comment (javitonino) [inline app/controllers/api/json/visualizations_controller.rb]: Ok to leaving these minor endpoints for a later PR.

File: app/controllers/carto/api/visualizations_controller.rb [hunk 2]
```diff
@@ -15,8 +18,9 @@ class VisualizationsController < ::Api::ApplicationController
       include Carto::UUIDHelper
       include Carto::ControllerHelper
       include VisualizationsControllerHelper
+      include Carto::VisualizationMigrator
 
-      ssl_required :index, :show, :destroy
+      ssl_required :index, :show, :create, :destroy
```
Comment (javitonino) [inline app/controllers/carto/api/visualizations_controller.rb]: Missing `update` here.

File: app/controllers/carto/api/visualizations_controller.rb [hunk 3]
```diff
@@ -124,6 +128,134 @@ def static_map
...
+        vis_data = payload
+
+        vis_data.delete(:permission)
+        vis_data.delete(:permission_id)
+
+        # Don't allow to modify next_id/prev_id, force to use set_next_id()
+        prev_id = vis_data.delete(:prev_id) || vis_data.delete('prev_id')
+        next_id = vis_data.delete(:next_id) || vis_data.delete('next_id')
+
+        param_tables = vis_data.delete(:tables)
```
Comment (javitonino) [inline app/controllers/carto/api/visualizations_controller.rb]: We mentioned the possibility of moving to a whitelist of parameters instead of this `delete` blacklisting. Did you try that?

File: app/controllers/carto/api/visualizations_controller.rb [hunk 4]
```diff
@@ -124,6 +128,134 @@ def static_map
...
+                return head(403) unless source && source.is_viewable_by_user?(user) && !source.kind_raster?
+                if source.derived?
+                  origin = 'copy'
+                  duplicate_derived_visualization(source_id, user)
+                else
+                  create_visualization_from_tables([source.user_table], vis_data)
+                end
+              elsif param_tables
+                subdomain = CartoDB.extract_subdomain(request)
+                viewed_user = ::User.find(username: subdomain)
```
Comment (javitonino) [inline app/controllers/carto/api/visualizations_controller.rb]: This can probably be `Carto::User`, since the table locator only relies on the user id.

File: app/controllers/carto/api/visualizations_controller.rb [hunk 5]
```diff
@@ -124,6 +128,134 @@ def static_map
...
+                tables = param_tables.map do |table_name|
+                  Carto::Helpers::TableLocator.new.get_by_id_or_name(table_name, viewed_user) if viewed_user
+                end
+                create_visualization_from_tables(tables.flatten, vis_data)
+              else
+                Carto::Visualization.new(vis_data.merge(name: name_candidate, user_id: current_user_id))
```
Comment (javitonino) [inline app/controllers/carto/api/visualizations_controller.rb]: FYI, I think this branch is not used at all. IIRC, when I refactored this controller (as part of the PR to make duplication use the visualization export feature), I saw that, but decided to leave this branch here just in case.

File: app/controllers/carto/api/visualizations_controller.rb [hunk 6]
```diff
@@ -124,6 +128,134 @@ def static_map
...
+              end
+
+        vis.ensure_valid_privacy
+        # both null, make sure is the first children or automatically link to the tail of the list
+        if !vis.parent_id.nil? && prev_id.nil? && next_id.nil?
```
Comment (javitonino) [inline app/controllers/carto/api/visualizations_controller.rb]: I'd take this chance to start removing Odyssey stuff (parent, child, next, prev). https://github.com/CartoDB/cartodb/issues/7677. There are more Odyssey-related changes (some of them, ugly) throughout the PR, that we could remove if we choose to start dropping support as of now.

File: app/controllers/carto/api/visualizations_controller.rb [hunk 7]
```diff
@@ -124,6 +128,134 @@ def static_map
...
+          parent_vis = Carto::Visualization.find(vis.parent_id)
+          return head(403) unless parent_vis.has_permission?(current_user, Carto::Permission::ACCESS_READWRITE)
+
+          children = parent_vis.children
+
+          prev_id = children.last.id unless children.empty?
+        end
+
+        vis.store
```
Comment (javitonino) [inline app/controllers/carto/api/visualizations_controller.rb]: This is just `vis.save!`. I'd change it as it's easier to understand (no need to check the `store` implementation). Same thing in `update`

File: app/controllers/carto/api/visualizations_controller.rb [hunk 8]
```diff
@@ -124,6 +128,134 @@ def static_map
...
+        render_jsonp(Carto::Api::VisualizationPresenter.new(vis, current_viewer, self).to_poro)
+      rescue => e
+        CartoDB::Logger.error(message: "Error creating visualization", visualization_id: vis.id, exception: e)
+        render_jsonp({ errors: vis.errors.full_messages }, 400)
+      end
+
+      def update
+        vis = @visualization
+
+        return head(404) unless vis
```
Comment (javitonino) [inline app/controllers/carto/api/visualizations_controller.rb]: This check is handled by `load_visualization` hook and can be removed.

File: app/controllers/carto/api/visualizations_controller.rb [hunk 9]
```diff
@@ -124,6 +128,134 @@ def static_map
...
+
+        vis.transition_options = params[:transition_options] if params[:transition_options]
+
+        # Don't allow to modify next_id/prev_id, force to use set_next_id()
+        vis_data.delete(:prev_id) || vis_data.delete('prev_id')
+        vis_data.delete(:next_id) || vis_data.delete('next_id')
+        # when a table gets renamed, its canonical visualization is renamed, so we must revert renaming if that failed
+        # This is far from perfect, but works without messing with table-vis sync and their two backends
+
+        valid_attributes = valid_update_visualization_attributes(vis_data)
```
Comment (javitonino) [inline app/controllers/carto/api/visualizations_controller.rb]: Since this does the `slice`, what about removing the `deletes` above it?

File: app/controllers/carto/api/visualizations_controller.rb [hunk 10]
```diff
@@ -124,6 +128,134 @@ def static_map
...
+            migrate_visualization_to_v3(vis)
+          end
+        end
+
+        # TODO: sometimes an attribute changes. Example: visualization name because of sanitization.
+        # Avoid this reload by updating the entity properly.
+        vis.reload
+
+        render_jsonp(Carto::Api::VisualizationPresenter.new(vis, current_viewer, self).to_poro)
+      rescue KeyError => e
```
Comment (javitonino) [inline app/controllers/carto/api/visualizations_controller.rb]: Pretty sure we can remove this `rescue`. It's the equivalent of `RecordNotFound` for `Member`

File: app/models/carto/visualization.rb [hunk 11]
```diff
@@ -554,6 +560,77 @@ def has_permission?(user, permission_type)
...
+    self.privacy = PRIVACY_PUBLIC unless can_be_private?
+  end
+
+  def default_privacy
+    can_be_private? ? PRIVACY_LINK : PRIVACY_PUBLIC
+  end
+
+  # @param other_vis CartoDB::Visualization::Member|nil
+  # Note: Changes state both of self, other_vis and other affected list items, but only reloads self & other_vis
+  def set_next_list_item!(other_vis)
```
Comment (javitonino) [inline app/models/carto/visualization.rb]: Ugh! If this doesn't convince you to remove Odyssey, nothing will.

File: app/models/carto/visualization.rb [hunk 12]
```diff
@@ -554,6 +560,77 @@ def has_permission?(user, permission_type)
...
+    end
+
+    store
+
+    # TODO: remove this reload
+    reload
+  end
+
+  def privacy=(privacy)
+    super(privacy ? privacy.downcase : nil)
```
Comment (javitonino) [inline app/models/carto/visualization.rb]: `super(privacy.try(:downcase))`?

File: app/controllers/carto/api/visualizations_controller.rb [hunk 13]
```diff
@@ -124,6 +128,134 @@ def static_map
...
+          if new_vis_name != old_vis_name && vis.table.name == old_table_name
+            vis.name = old_vis_name
+            vis.store
+          else
+            vis
+          end
+        else
+          old_version = vis.version
+
+          vis.attributes = valid_attributes
```
Comment (javitonino) [inline app/controllers/carto/api/visualizations_controller.rb]: This works, but looks dangerous. Maybe `vis.assign_attributes(valid_attributes)` it's better.

Looking at [the docs](https://apidock.com/rails/ActiveRecord/Base/attributes%3D), this seems to be actually fine, so feel free to ignore 🤐

File: app/controllers/api/json/visualizations_controller.rb (hunk 1 above)
Comment (juanignaciosl) [inline app/controllers/api/json/visualizations_controller.rb]: We could probably not only not delete watching but make it more powerful. We often have issues with people editing the same dataset/map from different tabs. But that's another discussion.

File: app/controllers/carto/api/visualizations_controller.rb (hunk 2 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/visualizations_controller.rb]: ¬¬U

File: app/controllers/carto/api/visualizations_controller.rb (hunk 3 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/visualizations_controller.rb]: Yes, it's `valid_update_visualization_attributes` method. I didn't think about removing those `delete`s. They're nice in the sense that they make explicit that something is not a visualization _attribute_ but a parameter that will be used in a different way. I think that whitelisting covers another concern. Do you think that replacing the deletion would be better?

File: app/controllers/carto/api/visualizations_controller.rb (hunk 5 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/visualizations_controller.rb]: Maybe it's not _currently_ used, but only taking a look at the controller you can't guarantee that, can you?

File: app/controllers/carto/api/visualizations_controller.rb (hunk 6 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/visualizations_controller.rb]: I just pinged the team, but it should be fine. I must admit that I spent some minutes making this work, but we should probably let this go.

File: app/controllers/carto/api/visualizations_controller.rb (hunk 5 above)
Comment (javitonino) [inline app/controllers/carto/api/visualizations_controller.rb]: No, of course not. I mean, it's not used from frontend (you never create a vsualization without layers).

On the other hand, I just saw that this doesn't create a map that I can see. So if this is called, it will create a visualization without map, that is sure to be broken.

But anyway, just wanted to mention since I did some research in the past. We don't have to necessarily remove this one now, in the context of the refactor.

File: app/controllers/carto/api/visualizations_controller.rb (hunk 7 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/visualizations_controller.rb]: Ok. In fact I didn't remember the added `def store` for backwards compatibility, I got surprised when I saw it, but then I forgot to remove it 👍

File: app/controllers/carto/api/visualizations_controller.rb (hunk 8 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/visualizations_controller.rb]: 👍

File: app/controllers/carto/api/visualizations_controller.rb (hunk 13 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/visualizations_controller.rb]: Yep, even [becoming an alias](https://apidock.com/rails/ActiveRecord/AttributeAssignment/attributes%3D).

File: app/models/carto/visualization.rb (hunk 11 above)
Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: 😆

File: app/controllers/carto/api/visualizations_controller.rb (hunk 10 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/visualizations_controller.rb]: 👍

File: app/controllers/carto/api/visualizations_controller.rb [hunk 14]
```diff
@@ -25,15 +29,16 @@ class VisualizationsController < ::Api::ApplicationController
...
+                                                :static_map, :vizjson2, :vizjson3, :update, :destroy]
       before_filter :ensure_visualization_owned, only: [:destroy]
 
       rescue_from Carto::LoadError, with: :rescue_from_carto_error
       rescue_from Carto::UUIDParameterFormatError, with: :rescue_from_carto_error
 
       def show
         render_jsonp(to_json(@visualization))
-      rescue KeyError
+      rescue => e
```
Comment (javitonino) [inline app/controllers/carto/api/visualizations_controller.rb]: Can this even be raised? I'd remove it if not. I'm mentioning this because we are losing information in this change. Prior to this change, this was supposed to catch visualization not found errors (which should be handled by a before filter now in this controller). If we do the change, we don't know what is this intending to catch (probably nothing).

Anyway, only remove it if you are confident, it's not important and we have a log in Rollbar to see what happened if this is triggered.

File: app/controllers/carto/api/visualizations_controller.rb [hunk 15]
```diff
@@ -236,6 +338,48 @@ def carto_referer?
       rescue URI::InvalidURIError
         false
       end
+
+      def payload
```
Comment (javitonino) [inline app/controllers/carto/api/visualizations_controller.rb]: Can you replace this with `params`? It would be nicer. I'd try it and abort if it starts to get complicated.

File: app/controllers/carto/api/visualizations_controller.rb [hunk 16]
```diff
@@ -236,6 +338,48 @@ def carto_referer?
...
+
+      def duplicate_derived_visualization(source, user)
+        export_service = Carto::VisualizationsExportService2.new
+        visualization_hash = export_service.export_visualization_json_hash(source, user)
+        visualization_copy = export_service.build_visualization_from_hash_export(visualization_hash)
+        visualization_copy.name = name_candidate
+        visualization_copy.version = user.new_visualizations_version
+        Carto::VisualizationsExportPersistenceService.new.save_import(user, visualization_copy)
+
+        Carto::Visualization.find(visualization_copy.id)
```
Comment (javitonino) [inline app/controllers/carto/api/visualizations_controller.rb]: This was a hack in the old controller to return a `::Member` for consistency with other code paths. Now that everything uses `Carto::Visualizations`, you should be able to just return `visualization_copy` here.

File: app/models/carto/visualization.rb [hunk 17]
```diff
@@ -178,6 +176,10 @@ def transition_options
     @transition_options ||= (slide_transition_options.nil? ? {} : JSON.parse(slide_transition_options).symbolize_keys)
   end
 
+  def transition_options=(value)
+    self.slide_transition_options = ::JSON.dump(value.nil? ? DEFAULT_OPTIONS_VALUE : value)
+  end
```
Comment (javitonino) [inline app/models/carto/visualization.rb]: This field is for Odyssey as well... 🔥

File: app/controllers/carto/api/visualizations_controller.rb (hunk 14 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/visualizations_controller.rb]: It's a _#show_, so it's really rare that this fails. I chose to keep the logging in order to have an explicit point of failure. Maybe it's redundant, but I think that it's harmless and might help with some rare errors.

File: app/controllers/carto/api/visualizations_controller.rb (hunk 15 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/visualizations_controller.rb]: I must admit that I don't know it. This seems quite intentional:

```ruby
      def payload
        request.body.rewind
        ::JSON.parse(request.body.read.to_s || String.new, symbolize_names: true)
      end
```

I guess that it was added for some reason, good or not. I'd prefer not removing this in this refactor, as it's already quite important.

**Comment by javitonino** [comment]
🔴 💥

**Comment by juanignaciosl** [comment]
Although I'd love to remove Odyssey support at model level, it's complicated at this moment. Removing `transition_options` makes old presenters and adapters incompatible, and vizJSON v2 would be affected. I don't think that a model refactor PR is the moment to make existing data unavailable. Even though UI doesn't support Odyssey anymore, a refactor shouldn't break existing vizJSONs, IMHO.

Let's discuss!

PS: reverting the deletion led to 🍏 , ready for CR.

**Comment by javitonino** [comment]
About odyssey: This PR is about visualization controller refactor, not removing Odyssey. I proposed starting the removal now because it was adding a lot of junk to the PR that we did not need. But this is not the time to complete the Odyssey removal.

So, I agree with the current approach. Remove what is helpful to remove (simplifies PR), and leave the problematic stuff for another time. Also, not the time to touch vizjson v2, just in case this breaks non-odyssey maps.

tl;dr: to acceptance!

**Comment by javitonino** [comment]
- New debug Rollbar trace: https://rollbar.com/vizzuality/CartoDB/items/23283 (when changing dataset privacy). Probably fine, I'll check later.
- Failure renaming dataset in organization https://rollbar.com/vizzuality/CartoDB/items/23284/ -> 🔴 Leaves DB in inconsistent state (renamed in psql, not in carto, ghost tables fixes it)
- Something strange happens sharing map with org member. In the receipient, it shows 1 shared map but it doesn't list it, I'll investigate, might be something in my staging org.

screenshot_20170601_143621

**Comment by javitonino** [comment]
Fixed 2. 1 seems fine. 3 seems like something broken in my staging. I'll retry in the next acceptance round.

**Comment by juanignaciosl** [comment]
Nice tests and changes, thanks!

**Comment by javitonino** [comment]
Found another one. Creating a map from dataset, in builder -> https://rollbar.com/vizzuality/CartoDB/items/23290/?item_page=0&item_count=100&#instances -> 🔴

**Comment by javitonino** [comment]
Found another one. Creating a map from dataset view, in builder -> https://rollbar.com/vizzuality/CartoDB/items/23290/?item_page=0&item_count=100&#instances -> 🔴

File: app/controllers/carto/api/visualizations_controller.rb [hunk 18]
```diff
@@ -262,6 +262,16 @@ def destroy
 
       private
 
+      # excluded:
+      #   :id, :map_id, :type, :created_at, :external_source, :url, :version, :likes, :liked, :table, :user_id
+      #   :synchronization, :uses_builder_features, :auth_tokens, :transition_options, :prev_id, :next_id, :parent_id
+      #   :active_child, :permission
+      VALID_UPDATE_ATTRIBUTES = [:name, :display_name, :active_layer_id, :tags, :description, :privacy, :updated_at,
```
Comment (mayoral) [inline app/controllers/carto/api/visualizations_controller.rb]: Nice use of constants to DRY. Also makes more clear what fields are added in the create vs the update. 👍
